R/scale-hue.r

```diff
@@ -63,3 +62,120 @@ scale_fill_hue <- function(..., h = c(0, 360) + 15, c = 100, l = 65, h.start = 0
+#' cty_by_var <- function(var) {
+#'   ggplot(mpg, aes(cty, colour = factor({{var}}), fill = factor({{var}}))) +
+#'     geom_density(alpha = 0.2)
+#' }
+#' # The default color scale for three levels
+#' cty_by_var(class)
+#'
+#' # Define custom palettes for when there are 1-2, 3, or 4-6 levels
+#' opts <- options(
+#'   ggplot2.discrete.fill = list(
+#'     c("skyblue", "orange"),
+#'     RColorBrewer::brewer.pal(3, "Set2"),
+#'     RColorBrewer::brewer.pal(6, "Accent")
+#'   )
+#' )
+#' cty_by_var(year)
+#' cty_by_var(drv)
+#' cty_by_var(fl)
+#' cty_by_var(class)
+#' options(opts)
+#'
+scale_colour_discrete <- function(..., type = getOption("ggplot2.discrete.colour", getOption("ggplot2.discrete.fill"))) {
+  type <- type %||% okabeIto
+  if (is.function(type)) {
+    type(...)
+  } else {
+    scale_colour_qualitative(..., codes = type)
+  }
+}
+
+#' @rdname scale_colour_discrete
+#' @export
+scale_fill_discrete <- function(..., type = getOption("ggplot2.discrete.fill", getOption("ggplot2.discrete.colour"))) {
+  type <- type %||% okabeIto
+  if (is.function(type)) {
+    type(...)
+  } else {
+    scale_fill_qualitative(..., codes = type)
+  }
+}
+
+# TODO: export?
+scale_colour_qualitative <- function(..., codes = NULL, h = c(0, 360) + 15, c = 100, l = 65, h.start = 0,
+                                     direction = 1, na.value = "grey50", aesthetics = "colour") {
+  discrete_scale(
+    aesthetics, "qualitative", qualitative_pal(codes, h, c, l, h.start, direction),
+    na.value = na.value, ...
+  )
+}
+
+scale_fill_qualitative <- function(..., codes = NULL, h = c(0, 360) + 15, c = 100, l = 65, h.start = 0,
+                                   direction = 1, na.value = "grey50", aesthetics = "fill") {
+  discrete_scale(
+    aesthetics, "qualitative", qualitative_pal(codes, h, c, l, h.start, direction),
+    na.value = na.value, ...
+  )
+}
+
+qualitative_pal <- function(codes, h, c, l, h.start, direction) {
+  function(n) {
+    if (!length(codes))  {
+      return(scales::hue_pal(h, c, l, h.start, direction)(n))
+    }
+    codes_list <- if (!is.list(codes)) list(codes) else codes
+    if (!all(vapply(codes_list, is.character, logical(1)))) {
+      stop("codes must be a character vector or a list of character vectors", call. = FALSE)
+    }
+    codes_lengths <- vapply(codes_list, length, integer(1))
+    # If there are more levels than color codes default to hue_pal()
+    if (max(codes_lengths) < n) {
+      return(scales::hue_pal(h, c, l, h.start, direction)(n))
+    }
+    # Use the minimum length vector that exceeds the number of levels (n)
+    codes_list <- codes_list[order(codes_lengths)]
+    i <- 1
+    while (length(codes_list[[i]]) < n) {
+      i <- i + 1
+    }
+    codes_list[[i]][seq_len(n)]
+  }
+}
+
+# prismatic::check_color_blindness(okabeIto)
+okabeIto <- c(
+  "#E69F00",
+  "#56B4E9",
+  "#009E73",
+  "#F0E442",
+  "#0072B2",
+  "#D55E00",
+  "#CC79A7",
+  "#999999"
```

Comment:
I've received feedback from colorblind people that gray is not that great of a choice and that black would be better. See also: https://github.com/jrnold/ggthemes/blob/master/data-raw/theme-data/colorblind.yml (And note black is the first color there. Not sure if I like it or not, but it definitely works better for people with CVD.)

Reply:
Interesting, thanks. I did (somewhat selfishly) avoid using black as I was hoping for a default that could work well in both 'light' and 'dark' mode (so I can use it for the [upcoming auto-theming feature in shiny](https://github.com/rstudio/shiny/pull/2740)). Now that I think of it though, maybe it'd be better to switch between black/white based on the background color?  It'd be easy enough for me to do that for my purposes, but maybe that's worth doing in ggplot2 itself (i.e., use white instead of black if `plot.background` is dark)

Reply:
This touches on something that I have said for a long time: I think color scales should be settable from the theme. That makes much more sense to me than setting them via global options.

Reply:
See #2691.

Reply:
Can we agree to use black for now, and improve the support for dark theme after #2691 gets ready?

Reply:
ok with me